src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -276,116 +308,160 @@ starting_address(h::HubbardRealSpace) = h.address
 
 dimension(::HubbardRealSpace, address) = number_conserving_dimension(address)
 
-function diagonal_element(h::HubbardRealSpace, address)
-    int = isnothing(h.u) ? 0.0 : local_interaction(address, h.u)
-    pot = isnothing(h.v) ? 0.0 : external_potential(address, h.potential)
-    return int + pot
-end
-function diagonal_element(h::HubbardRealSpace{1}, address)
-    int = isnothing(h.u) ? 0.0 : local_interaction(address, h.u[1])
-    pot = if isnothing(h.v)
-            0.0
-        else
-            @boundscheck checkbounds(h.potential, 1:num_modes(address), 1)
-            @inbounds external_potential(address, @view h.potential[:,1])
-        end
-    return int + pot
-end
-
-###
-### Offdiagonals
-###
-# This may be an inefficient implementation, but it is not actually used anywhere in the
-# main algorithm.
-get_offdiagonal(h::HubbardRealSpace, add, i) = offdiagonals(h, add)[i]
-num_offdiagonals(h::HubbardRealSpace, add) = length(offdiagonals(h, add))
-
-"""
-    HubbardRealSpaceCompOffdiagonals{G,A} <: AbstractOffdiagonals{A,Float64}
-
-Offdiagonals for a single address component. Used with [`HubbardRealSpace`](@ref) model
-with a single-component address, or a component of a [`CompositeFS`](@ref).
-"""
-struct HubbardRealSpaceCompOffdiagonals{G,A} <: AbstractOffdiagonals{A,Float64}
+# offdiaonals =========================================================================== #
+# Holds the offdiagonals for a single-component nearest neighbour one-body term. It's
+# structured like a matric where the first index determines the occupied site in the adress
+# and the second index determines the site the particle will hop to.
+struct HubbardRealSpaceComponentData{I,G,A,C,O} <: AbstractMatrix{Pair{A,Float64}}
```

Comment:
It looks like this supports only real valued hoppings. This may be OK for now, but it could be useful to generalise this to complex values later.